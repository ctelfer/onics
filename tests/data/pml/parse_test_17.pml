# Braceless if and while statements
#

BEGIN {
	x = 0;
	y = 1;

	# simple one stmt 'if'
	if ( x )
		y = 2;

	# simple one stmt with 'print' as stmt
	if ( x == y )
		print "equal";

	# simple one stmt 'if-else'
	if ( y )
		x = 3;
	else
		y = 10;

	# simple one stmt 'while'
	while ( x < 5 )
		x = x + 1;

	# while with if-else for stmt
	while ( x < 10 )
		if ( x == 2 )
			x = 5;
		else
			x = x + 1;

	# stmtlist for 'if', stmt for 'else'
	if ( x ) {
		print "hi";
		x = 2;
	} else
		y = 3;

	# stmt for if and stmtlist for 'else'
	if ( y == 10 )
		z = 1;
	else {
		x = y * z;
		print "hello";
	}

	# dangling else
	if ( x ) if ( y ) x = 5; else y = 10;
}
