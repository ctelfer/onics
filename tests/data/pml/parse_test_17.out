#########
Initializing parse data structures
#########
#########
Starting Parse
#########


########
Done parsing, destroying scanner and parser
########


########
Optimizing tree:
########
Done... printing optimized tree
Printing PML Abstract Syntax Tree
-----------
Variables
-----------
   Variable: mem [global; width=0, addr=0]
-----------
Functions
-----------
   Function[intrinsic]: str_len() -- 1 args, 0-byte stack, return type=scalar
   Parameters & Variables -- 
      Variable: ref [param; width=16, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: str_addr() -- 1 args, 0-byte stack, return type=scalar
   Parameters & Variables -- 
      Variable: ref [param; width=16, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: str_ispkt() -- 1 args, 0-byte stack, return type=scalar
   Parameters & Variables -- 
      Variable: ref [param; width=16, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: str_seg() -- 1 args, 0-byte stack, return type=scalar
   Parameters & Variables -- 
      Variable: ref [param; width=16, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: str_isnull() -- 1 args, 0-byte stack, return type=scalar
   Parameters & Variables -- 
      Variable: ref [param; width=16, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: str_mkref() -- 4 args, 0-byte stack, return type=string reference
   Parameters & Variables -- 
      Variable: ispkt [param; width=4, addr=0]
      Variable: seg [param; width=4, addr=1]
      Variable: addr [param; width=4, addr=2]
      Variable: len [param; width=4, addr=3]
   Body -- 
      (null)
   Function[intrinsic]: pkt_new() -- 2 args, 0-byte stack, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=4, addr=0]
      Variable: len [param; width=4, addr=1]
   Body -- 
      (null)
   Function[intrinsic]: pkt_new_z() -- 2 args, 0-byte stack, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=4, addr=0]
      Variable: len [param; width=4, addr=1]
   Body -- 
      (null)
   Function[intrinsic]: pkt_swap() -- 2 args, 0-byte stack, return type=void
   Parameters & Variables -- 
      Variable: pndst [param; width=4, addr=0]
      Variable: pnsrc [param; width=4, addr=1]
   Body -- 
      (null)
   Function[intrinsic]: pkt_copy() -- 2 args, 0-byte stack, return type=void
   Parameters & Variables -- 
      Variable: pndst [param; width=4, addr=0]
      Variable: pnsrc [param; width=4, addr=1]
   Body -- 
      (null)
   Function[intrinsic]: pkt_ins_u() -- 3 args, 0-byte stack, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=4, addr=0]
      Variable: off [param; width=4, addr=1]
      Variable: len [param; width=4, addr=2]
   Body -- 
      (null)
   Function[intrinsic]: pkt_ins_d() -- 3 args, 0-byte stack, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=4, addr=0]
      Variable: off [param; width=4, addr=1]
      Variable: len [param; width=4, addr=2]
   Body -- 
      (null)
   Function[intrinsic]: pkt_cut_u() -- 1 args, 0-byte stack, return type=void
   Parameters & Variables -- 
      Variable: str [param; width=16, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: pkt_cut_d() -- 1 args, 0-byte stack, return type=void
   Parameters & Variables -- 
      Variable: str [param; width=16, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: pkt_parse() -- 1 args, 0-byte stack, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=4, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: pkt_get_off() -- 4 args, 0-byte stack, return type=scalar
   Parameters & Variables -- 
      Variable: pnum [param; width=4, addr=0]
      Variable: prid [param; width=4, addr=1]
      Variable: idx [param; width=4, addr=2]
      Variable: oid [param; width=4, addr=3]
   Body -- 
      (null)
   Function[intrinsic]: pkt_adj_off() -- 5 args, 0-byte stack, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=4, addr=0]
      Variable: prid [param; width=4, addr=1]
      Variable: idx [param; width=4, addr=2]
      Variable: oid [param; width=4, addr=3]
      Variable: amt [param; width=4, addr=4]
   Body -- 
      (null)
   Function[intrinsic]: parse_push_back() -- 2 args, 0-byte stack, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=4, addr=0]
      Variable: prid [param; width=4, addr=1]
   Body -- 
      (null)
   Function[intrinsic]: parse_pop_back() -- 1 args, 0-byte stack, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=4, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: parse_push_front() -- 2 args, 0-byte stack, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=4, addr=0]
      Variable: prid [param; width=4, addr=1]
   Body -- 
      (null)
   Function[intrinsic]: parse_pop_front() -- 1 args, 0-byte stack, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=4, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: parse_update() -- 1 args, 0-byte stack, return type=void
   Parameters & Variables -- 
      Variable: pdesc [param; width=4, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: fix_dltype() -- 1 args, 0-byte stack, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=4, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: fix_len() -- 1 args, 0-byte stack, return type=void
   Parameters & Variables -- 
      Variable: pdesc [param; width=4, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: fix_all_len() -- 1 args, 0-byte stack, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=4, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: fix_csum() -- 1 args, 0-byte stack, return type=void
   Parameters & Variables -- 
      Variable: pdesc [param; width=4, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: fix_all_csum() -- 1 args, 0-byte stack, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=4, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: meta_get_ts_sec() -- 1 args, 0-byte stack, return type=scalar
   Parameters & Variables -- 
      Variable: pnum [param; width=4, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: meta_get_ts_nsec() -- 1 args, 0-byte stack, return type=scalar
   Parameters & Variables -- 
      Variable: pnum [param; width=4, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: meta_set_ts() -- 3 args, 0-byte stack, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=4, addr=0]
      Variable: sec [param; width=4, addr=1]
      Variable: nsec [param; width=4, addr=2]
   Body -- 
      (null)
   Function[intrinsic]: meta_get_presnap() -- 1 args, 0-byte stack, return type=scalar
   Parameters & Variables -- 
      Variable: pnum [param; width=4, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: meta_set_presnap() -- 2 args, 0-byte stack, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=4, addr=0]
      Variable: val [param; width=4, addr=1]
   Body -- 
      (null)
   Function[intrinsic]: meta_get_inport() -- 1 args, 0-byte stack, return type=scalar
   Parameters & Variables -- 
      Variable: pnum [param; width=4, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: meta_set_inport() -- 2 args, 0-byte stack, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=4, addr=0]
      Variable: val [param; width=4, addr=1]
   Body -- 
      (null)
   Function[intrinsic]: meta_get_outport() -- 1 args, 0-byte stack, return type=scalar
   Parameters & Variables -- 
      Variable: pnum [param; width=4, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: meta_set_outport() -- 2 args, 0-byte stack, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=4, addr=0]
      Variable: val [param; width=4, addr=1]
   Body -- 
      (null)
   Function[intrinsic]: meta_get_flowid() -- 1 args, 0-byte stack, return type=scalar
   Parameters & Variables -- 
      Variable: pnum [param; width=4, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: meta_set_flowid() -- 2 args, 0-byte stack, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=4, addr=0]
      Variable: val [param; width=4, addr=1]
   Body -- 
      (null)
   Function[intrinsic]: meta_get_class() -- 1 args, 0-byte stack, return type=scalar
   Parameters & Variables -- 
      Variable: pnum [param; width=4, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: meta_set_class() -- 2 args, 0-byte stack, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=4, addr=0]
      Variable: val [param; width=4, addr=1]
   Body -- 
      (null)
   Function[intrinsic]: meta_get_seq() -- 1 args, 0-byte stack, return type=scalar
   Parameters & Variables -- 
      Variable: pnum [param; width=4, addr=0]
   Body -- 
      (null)
   Function[intrinsic]: meta_set_seq() -- 2 args, 0-byte stack, return type=void
   Parameters & Variables -- 
      Variable: pnum [param; width=4, addr=0]
      Variable: val [param; width=4, addr=1]
   Body -- 
      (null)
   Function[intrinsic]: exit() -- 1 args, 0-byte stack, return type=void
   Parameters & Variables -- 
      Variable: rval [param; width=4, addr=0]
   Body -- 
      (null)
   Function[inline, intrinsic, pconst]: pop() -- 1 args, 0-byte stack, return type=scalar
   Parameters & Variables -- 
      Variable: num [param; width=4, addr=0]
   Body -- 
      (null)
   Function[inline, intrinsic, pconst]: log2() -- 1 args, 0-byte stack, return type=scalar
   Parameters & Variables -- 
      Variable: num [param; width=4, addr=0]
   Body -- 
      (null)
   Function[inline, intrinsic, pconst]: min() -- 2 args, 0-byte stack, return type=scalar
   Parameters & Variables -- 
      Variable: num1 [param; width=4, addr=0]
      Variable: num2 [param; width=4, addr=1]
   Body -- 
      (null)
   Function[inline, intrinsic, pconst]: max() -- 2 args, 0-byte stack, return type=scalar
   Parameters & Variables -- 
      Variable: num1 [param; width=4, addr=0]
      Variable: num2 [param; width=4, addr=1]
   Body -- 
      (null)
-----------
Begin Rule
-----------
   Rule: (begin)
      Action Variables -- 
      Variable: x [local; width=4, addr=0]
      Variable: y [local; width=4, addr=1]
      Variable: z [local; width=4, addr=2]
   Action -- 
      List:
      -----
      Assignment to x
      Value -- 
         Scalar [[c]; width=4] -- width 4: 0 (0,0x0)
      -----
      Assignment to y
      Value -- 
         Scalar [[c]; width=4] -- width 4: 1 (1,0x1)
      -----
      If Statement
      Test -- 
         Locator: variable(x) [[]; width=4]
         Variable -- 
            Variable: x [local; width=4, addr=0]
      True body -- 
         List:
         -----
         Assignment to y
         Value -- 
            Scalar [[c]; width=4] -- width 4: 2 (2,0x2)
         -----
      -----
      If Statement
      Test -- 
         Binary Operation: 'equals' [[]; width=4]
         Left Operand -- 
            Locator: variable(x) [[]; width=4]
            Variable -- 
               Variable: x [local; width=4, addr=0]
         Right Operand -- 
            Locator: variable(y) [[]; width=4]
            Variable -- 
               Variable: y [local; width=4, addr=1]
      True body -- 
         List:
         -----
         Print Statement: type: string, width:0, flags:[]
         Expression -- 
            Byte string [[c]; width=0] -- 
            6 bytes in read-only segment at address 0
            657175616c0a
         -----
      -----
      If Statement
      Test -- 
         Locator: variable(y) [[]; width=4]
         Variable -- 
            Variable: y [local; width=4, addr=1]
      True body -- 
         List:
         -----
         Assignment to x
         Value -- 
            Scalar [[c]; width=4] -- width 4: 3 (3,0x3)
         -----
      False body -- 
         List:
         -----
         Assignment to y
         Value -- 
            Scalar [[c]; width=4] -- width 4: 10 (10,0xa)
         -----
      -----
      While Statement
      Loop Test -- 
         Binary Operation: 'less than' [[]; width=4]
         Left Operand -- 
            Locator: variable(x) [[]; width=4]
            Variable -- 
               Variable: x [local; width=4, addr=0]
         Right Operand -- 
            Scalar [[c]; width=4] -- width 4: 5 (5,0x5)
      Loop Body -- 
         List:
         -----
         Assignment to x
         Value -- 
            Binary Operation: 'add' [[]; width=4]
            Left Operand -- 
               Locator: variable(x) [[]; width=4]
               Variable -- 
                  Variable: x [local; width=4, addr=0]
            Right Operand -- 
               Scalar [[c]; width=4] -- width 4: 1 (1,0x1)
         -----
      -----
      While Statement
      Loop Test -- 
         Binary Operation: 'less than' [[]; width=4]
         Left Operand -- 
            Locator: variable(x) [[]; width=4]
            Variable -- 
               Variable: x [local; width=4, addr=0]
         Right Operand -- 
            Scalar [[c]; width=4] -- width 4: 10 (10,0xa)
      Loop Body -- 
         List:
         -----
         If Statement
         Test -- 
            Binary Operation: 'equals' [[]; width=4]
            Left Operand -- 
               Locator: variable(x) [[]; width=4]
               Variable -- 
                  Variable: x [local; width=4, addr=0]
            Right Operand -- 
               Scalar [[c]; width=4] -- width 4: 2 (2,0x2)
         True body -- 
            List:
            -----
            Assignment to x
            Value -- 
               Scalar [[c]; width=4] -- width 4: 5 (5,0x5)
            -----
         False body -- 
            List:
            -----
            Assignment to x
            Value -- 
               Binary Operation: 'add' [[]; width=4]
               Left Operand -- 
                  Locator: variable(x) [[]; width=4]
                  Variable -- 
                     Variable: x [local; width=4, addr=0]
               Right Operand -- 
                  Scalar [[c]; width=4] -- width 4: 1 (1,0x1)
            -----
         -----
      -----
      If Statement
      Test -- 
         Locator: variable(x) [[]; width=4]
         Variable -- 
            Variable: x [local; width=4, addr=0]
      True body -- 
         List:
         -----
         Print Statement: type: string, width:0, flags:[]
         Expression -- 
            Byte string [[c]; width=0] -- 
            3 bytes in read-only segment at address 7
            68690a
         -----
         Assignment to x
         Value -- 
            Scalar [[c]; width=4] -- width 4: 2 (2,0x2)
         -----
      False body -- 
         List:
         -----
         Assignment to y
         Value -- 
            Scalar [[c]; width=4] -- width 4: 3 (3,0x3)
         -----
      -----
      If Statement
      Test -- 
         Binary Operation: 'equals' [[]; width=4]
         Left Operand -- 
            Locator: variable(y) [[]; width=4]
            Variable -- 
               Variable: y [local; width=4, addr=1]
         Right Operand -- 
            Scalar [[c]; width=4] -- width 4: 10 (10,0xa)
      True body -- 
         List:
         -----
         Assignment to z
         Value -- 
            Scalar [[c]; width=4] -- width 4: 1 (1,0x1)
         -----
      False body -- 
         List:
         -----
         Assignment to x
         Value -- 
            Binary Operation: 'multiply' [[]; width=4]
            Left Operand -- 
               Locator: variable(y) [[]; width=4]
               Variable -- 
                  Variable: y [local; width=4, addr=1]
            Right Operand -- 
               Locator: variable(z) [[]; width=4]
               Variable -- 
                  Variable: z [local; width=4, addr=2]
         -----
         Print Statement: type: string, width:0, flags:[]
         Expression -- 
            Byte string [[c]; width=0] -- 
            6 bytes in read-only segment at address 11
            68656c6c6f0a
         -----
      -----
      If Statement
      Test -- 
         Locator: variable(x) [[]; width=4]
         Variable -- 
            Variable: x [local; width=4, addr=0]
      True body -- 
         List:
         -----
         If Statement
         Test -- 
            Locator: variable(y) [[]; width=4]
            Variable -- 
               Variable: y [local; width=4, addr=1]
         True body -- 
            List:
            -----
            Assignment to x
            Value -- 
               Scalar [[c]; width=4] -- width 4: 5 (5,0x5)
            -----
         False body -- 
            List:
            -----
            Assignment to y
            Value -- 
               Scalar [[c]; width=4] -- width 4: 10 (10,0xa)
            -----
         -----
      -----
-----------
Packet Rules
-----------
-----------
End Rule
-----------
   (null)
-----------


########
Clearing tree:
########
